src/main/scala/io/qbeast/spark/internal/sources/QbeastDataSource.scala

```diff
@@ -47,39 +47,35 @@ class QbeastDataSource private[sources] (private val tableFactory: IndexedTableF
 
   override def shortName(): String = "qbeast"
 
-  override def inferSchema(options: CaseInsensitiveStringMap): StructType = StructType(Seq())
+  override def inferSchema(options: CaseInsensitiveStringMap): StructType = {
+    val parameters = options.asScala.toMap
+    val tableId = QbeastOptions.loadTableIDFromParameters(parameters)
+    QbeastContext.metadataManager.loadCurrentSchema(tableId)
+  }
 
   // Used to get the table of an existing one
   override def getTable(
       schema: StructType,
       partitioning: Array[Transform],
       properties: util.Map[String, String]): Table = {
-    val tableId = QbeastOptions.loadTableIDFromParameters(properties.asScala.toMap)
+    val tableProperties = new util.HashMap[String, String](properties).asScala
```

Comment:
why is it necessary to do all these transformations on this map? From java Map, to HashMap, to Scala Map, and then a few line down other toMap? Can't it just be converted once?

Reply:
Unfortunately yes. The `util.Map` passed as a parameter turned out to be immutable (`java.util.collections.Collections.unmodifiableMap()`), and `asScala` does not copy the values it is just a wrapper, so without making a copy, the modification below would fail.